apps/CA.pl.in

```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
         copy_pemfile($FILE,"${CATOP}/$CACERT", "CERTIFICATE");
     } else {
         print "Making CA certificate ...\n";
-        $RET = run("$REQ -new -keyout"
+        $RET = run("$REQ $OPENSSL_CONFIG -new -keyout"
                 . " ${CATOP}/private/$CAKEY"
                 . " -out ${CATOP}/$CAREQ");
-        $RET = run("$CA -create_serial"
+        $RET = run("$CA $OPENSSL_CONFIG -create_serial"
                 . " -out ${CATOP}/$CACERT $CADAYS -batch"
                 . " -keyfile ${CATOP}/private/$CAKEY -selfsign"
                 . " -extensions v3_ca"
                 . " -infiles ${CATOP}/$CAREQ") if $RET == 0;
         print "CA certificate is in ${CATOP}/$CACERT\n" if $RET == 0;
     }
-} elsif ($WHAT eq '-pkcs12' ) {
-    my $cname = $ARGV[1];
-    $cname = "My Certificate" unless defined $cname;
+    $RET = run("$X509 -in ${CATOP}/$CACERT -inform pem"
+            . " -out ${CATOP}/$CACERTDER -outform der") if $RET == 0;
+    print "DER format version is in ${CATOP}/$CACERTDER\n" if $RET == 0;
+}
+
+# Create a PKCS#12 file from the signed cert and key
+sub pkcs12
+{
+    my $cname = shift;
+    $cname = "$CNAMETXT" unless $cname;
     $RET = run("$PKCS12 -in $NEWCERT -inkey $NEWKEY"
             . " -certfile ${CATOP}/$CACERT"
-            . " -out $NEWP12"
+            . " -out $NEWP12 $PASSOPT"
             . " -export -name \"$cname\"");
-    print "PKCS #12 file is in $NEWP12\n" if $RET == 0;
-} elsif ($WHAT eq '-xsign' ) {
-    $RET = run("$CA -policy policy_anything -infiles $NEWREQ");
-} elsif ($WHAT eq '-sign' ) {
-    $RET = run("$CA -policy policy_anything -out $NEWCERT -infiles $NEWREQ");
+    print "PKCS#12 file is in $NEWP12\n" if $RET == 0;
+}
+
+# Sign a certificate based on a certificate request
+sub xsign
+{
+    my ($days,$day) = @_;
+    $days = "" unless $days;
+    if ($days eq '-days' and $day) {
+        $days = $days. " ". $day;
+    } elsif ($days !~ /^-days[[:blank:]]+\w+/) {
+        $days = "";
+    }
```

Comment:
It is not generally good to believe user input though this is not a web cgi. I don't want to let arbitrary number of arguments injected into sub routines. Anyway, excessive arguments must be discarded somewhere through. I do it right at the entrance.
In addition to it, when #1797 `-extra` extension is ported also into this milestone, circumstance may change. I think this style is considerably the best for now. I am going to examine how they looks like with #1797 later.

**About '-days xxx' argument handling:**
No, it is not so simple. `-days` arguments come into these sub routines via two way.
1. As arguments to CA.pl
2. Called internally from other subs or main scope
In case 1, a user will execute `CA.pl -xsign -days 730`, my `$days` gets '`-days`' and `$day` '`730`'. There is a chance he might unusually enclose the whole arguments in a pair of qtuotes, in that case `$days` gets "`-days 730`" and `$day` undef.
In case 2, we almost always pass `-days` arguments in a form of a variable `$DAYS`. Result is the same as the latter in case 1. So they converge onto three patterns:
1. $days="-days" $day="_xxx_"
2. $days="-days" $day=undef (erroneously)
3. $days="-days (extra spaces could be) _xxx_" $day=undef
I tested them all and came to this conclusion. I took just the `join()` style.
Just for record, these subs with `-days` expect two arguments **at most**. I don't see a reason to pass whole `@ARGV` to them.